### ISSUE TITLE
Batch sizes

### ISSUE THREAD

**Issue by ryanleary** [SeanNaren/deepspeech.pytorch#58]
I'm trying to get some insight into batch sizes and whether or not the performance I'm seeing is expected. It seems that I can't set batch sizes much more than say, 32, w/ my dual Titan Xs. It's further my understanding that `dataparallel` will split that batch of 32 across the two GPUs for an effective batch size of 16 per gpu per batch. The model I'm training is all default: 4 LSTM layers w/ 400 hidden units. Now this is a fair amount different than many of the DeepSpeech 2 configurations in the paper, but I am seeing references to them having batch sizes of 512 spread over 8 Titan X's. This implies that whatever system they're running allows them to support batches of 64 per gpu. Seems to me we should be able to get closer to this number unless I'm missing something. Any thoughts?

**Comment by SeanNaren** [comment]
Baidus' internal code is much more memory optimised than Torch (atleast for RNNs), allowing them to run much larger batch sizes per GPU. This has always been a crux of using RNNs! I'm honestly not sure where to exactly look to handle this, but since you have pascal cards [this](https://github.com/pytorch/pytorch/pull/1141) PR is something to look at. Persistent RNNs speed up RNNs a lot when using smaller batch sizes because there is less transfer of memory!

**Comment by ryanleary** [comment]
I've seen many references to their work on being efficient with memory. I think I'm just surprised that it's making a ~4x difference in batch size. Impressive (and good for them) I suppose. Will close if this behavior is roughly consistent with others are seeing.

**Comment by EgorLakomkin** [comment]
Yes, I have observed the same behaviour

**Comment by SeanNaren** [comment]
I'm going to tie #55 to this, with successful FP16 we should be allowed to double batch sizes (with slight performance loss due to cublas doing some funky stuff with pseudo FP16 that I was notified about)

**Comment by SeanNaren** [comment]
As explained [here](https://github.com/SeanNaren/deepspeech.pytorch/issues/55#issuecomment-303385491), sadly FP16 won't be viable for some time (atleast for training).

**Comment by ryanleary** [comment]
We filter the data used to under a certain length. If you look at the librispeech.py script, for example, you can see that there is a flag there for doing filtering when the manifest is created.

**Comment by haquynh1505** [comment]
@ryanleary : Ok I see. Thank you so much!
